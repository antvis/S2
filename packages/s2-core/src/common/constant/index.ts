export * from './events';
export * from './basic';
export * from './classnames';
export * from './condition';
export * from './dataConfig';
export * from './frozen';
export * from './interaction';
export * from './options';
export * from './scroll';
export * from './theme';
export * from './tooltip';
export * from './resize';
