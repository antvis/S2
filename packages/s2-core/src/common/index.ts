export * from './constant';
export * from './debug';
export * from './i18n';
export * from './icons';
export * from './interface';
export * from './store';
