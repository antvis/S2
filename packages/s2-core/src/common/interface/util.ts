export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
