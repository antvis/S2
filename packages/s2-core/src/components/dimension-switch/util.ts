import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
