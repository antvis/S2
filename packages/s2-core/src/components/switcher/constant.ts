import { ColIcon, RowIcon, ValueIcon } from '../icons/index';

export enum FieldType {
  Rows = 'rows',
  Cols = 'cols',
  Values = 'values',
}

export const SWITCHER_CONFIG = {
  [FieldType.Rows]: {
    text: '行头',
    icon: RowIcon,
  },
  [FieldType.Cols]: {
    text: '列头',
    icon: ColIcon,
  },
  [FieldType.Values]: {
    text: '值',
    icon: ValueIcon,
  },
} as const;
