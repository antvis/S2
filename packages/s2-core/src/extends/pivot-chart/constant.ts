import type { G2Spec } from '@antv/g2';
import { type S2DataConfig, type S2Options } from '@antv/s2';
import { ChartDataCell } from './cell/chart-data-cell';
import { AxisRowColumnClick } from './interaction/axis-click';
import { AxisHover } from './interaction/axis-hover';

export const DEFAULT_G2_SPEC: G2Spec = {
  type: 'interval',
  autoFit: true,
  animate: false,
  axis: false,
  legend: false,
};

export const DEFAULT_G2_AXIS_SPEC: G2Spec = {
  autoFit: true,
  animate: false,
  line: false,
  tick: false,
  grid: false,
};

export const FIXED_OPTIONS: S2Options = {
  hierarchyType: 'grid',
  interaction: {
    selectedCellsSpotlight: false,
    copy: {
      enable: false,
    },
  },
  style: {
    colCell: {
      hideValue: false,
    },
  },
};

export const DEFAULT_OPTIONS: S2Options = {
  chartCoordinate: 'cartesian',
  chartSpec: DEFAULT_G2_SPEC,
  dataCell: (viewMeta, spreadsheet) => new ChartDataCell(viewMeta, spreadsheet),
  interaction: {
    customInteractions: [
      {
        key: 'axisHover',
        interaction: AxisHover,
      },
      {
        key: 'axisClick',
        interaction: AxisRowColumnClick,
      },
    ],
  },
};

export const FIXED_DATA_CONFIG: Partial<S2DataConfig> = {
  fields: {
    customValueOrder: null,
  },
};

export const DEFAULT_MEASURE_SIZE = 200;
export const DEFAULT_ROW_AXIS_SIZE = 100;
export const DEFAULT_COL_AXIS_SIZE = 50;
export const DEFAULT_DIMENSION_SIZE = 50;

/**
 * row axis
 */
export const KEY_GROUP_ROW_AXIS_SCROLL = 'rowAxisScrollGroup';
export const KEY_GROUP_ROW_AXIS_FROZEN = 'rowAxisHeaderFrozenGroup';
export const KEY_GROUP_ROW_AXIS_HEADER_FROZEN_TRAILING =
  'rowAxisHeaderFrozenTrailingGroup';
export const KEY_GROUP_ROW_AXIS_RESIZE_AREA = 'rowAxisHeaderResizeArea';

/**
 * column axis
 */
export const KEY_GROUP_COL_AXIS_SCROLL = 'colAxisScrollGroup';
export const KEY_GROUP_COL_AXIS_FROZEN = 'colAxisFrozenGroup';
export const KEY_GROUP_COL_AXIS_FROZEN_TRAILING = 'colAxisFrozenTrailingGroup';
export const KEY_GROUP_COL_AXIS_RESIZE_AREA = 'colAxisHeaderResizeArea';

export const PLACEHOLDER_FIELD = '$$placeholder$$';
