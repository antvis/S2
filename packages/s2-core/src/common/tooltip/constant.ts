import { PRE_CLASS } from '@/common/constant/classnames';

export const TOOLTIP_CLASS_PRE = `${PRE_CLASS}-tooltip`;

export const TOOLTIP_OPERATION_CLASS_PRE = `${TOOLTIP_CLASS_PRE}-operation`;

export const DEFAULT_ICON_PROPS = {
  width: 14,
  height: 14,
  style: {
    verticalAlign: 'sub',
    marginRight: 4,
  },
};

export const POSITION_X_OFFSET = 10;
export const POSITION_Y_OFFSET = 10;
