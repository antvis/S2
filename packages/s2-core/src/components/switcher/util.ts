import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
