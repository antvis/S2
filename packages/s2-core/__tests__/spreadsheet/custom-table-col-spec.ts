import type { Group } from '@antv/g';
import { getContainer } from 'tests/util/helpers';
import { waitForRender } from 'tests/util';
import { KEY_GROUP_COL_RESIZE_AREA } from '../../src/common/constant';
import {
  customColMultipleColumns,
  customColSimpleColumns,
} from '../data/custom-table-col-fields';
import { data } from '../data/mock-dataset.json';
import {
  expectHighlightActiveNodes,
  mapCellNodeValues,
} from '../util/interaction';
import { SpreadSheet, TableSheet } from '@/sheet-type';
import type { S2DataConfig, S2Options } from '@/common/interface';
import type { CustomRect } from '@/engine';

const s2Options: S2Options = {
  width: 600,
  height: 400,
};

describe('TableSheet Custom Tests', () => {
  let s2: SpreadSheet;

  const baseDataConfig: Pick<S2DataConfig, 'data' | 'meta'> = {
    data,
    meta: [
      {
        field: 'type',
        name: '类型',
      },
      {
        field: 'sub_type',
        name: '子类型',
      },
      {
        field: 'a-1',
        name: '层级1',
      },
      {
        field: 'a-1-1',
        name: '层级2',
      },
      {
        field: 'measure-1',
        name: '层级3',
      },
    ],
  };

  const customColDataCfg: S2DataConfig = {
    ...baseDataConfig,
    fields: {
      columns: customColSimpleColumns,
    },
  };

  beforeEach(async () => {
    s2 = new TableSheet(getContainer(), customColDataCfg, s2Options);
    await s2.render();
  });

  afterEach(() => {
    s2.destroy();
  });

  test('should render custom layout column nodes', () => {
    const colNodes = s2.facet.getColNodes().map((node) => {
      return {
        value: node.value,
        width: node.width,
        height: node.height,
        description: node.extra?.['description'],
      };
    });

    expect(colNodes).toHaveLength(7);
    expect(colNodes).toMatchSnapshot();
  });

  test('should calc correctly col index of leaf nodes', () => {
    const colLeafNodes = s2.facet.getColLeafNodes().map((node) => {
      return {
        value: node.value,
        colIndex: node.colIndex,
      };
    });

    expect(colLeafNodes).toHaveLength(5);
    expect(colLeafNodes).toMatchSnapshot();
  });

  test('should calc correctly leaf nodes width after column resized', async () => {
    s2.setOptions({
      style: {
        colCell: {
          heightByField: {
            area: 100,
          },
        },
      },
    });
    await s2.render(false);

    const colNodes = s2.facet.getColNodes().map((node) => {
      return {
        value: node.value,
        height: node.height,
      };
    });

    expect(colNodes).toMatchSnapshot();
  });

  test('should select custom col header cell', () => {
    // 地区
    const colNode = s2.facet.getColNodes()[0];

    // 选中地区
    s2.interaction.selectHeaderCell({
      cell: colNode.belongsCell!,
    });

    // 选中单元格本身
    expect(s2.interaction.getActiveCells()).toHaveLength(1);
    // 高亮子节点
    expectHighlightActiveNodes(s2, [
      'root[&]area[&]province',
      'root[&]area[&]city',
    ]);

    // 取消选中 a - 1
    s2.interaction.selectHeaderCell({
      cell: colNode.belongsCell!,
    });
    expect(s2.interaction.getActiveCells()).toBeEmpty();
  });

  test('should hide columns', async () => {
    const hiddenColumns = ['root[&]money[&]price'];

    await waitForRender(s2, () => {
      s2.interaction.hideColumns(hiddenColumns);
    });

    const hiddenColumnsDetail = s2.store.get('hiddenColumnsDetail', []);
    const [measureDetail] = hiddenColumnsDetail;

    expect(s2.options.interaction?.hiddenColumnFields).toEqual(hiddenColumns);
    expect(s2.facet.getColNodes().map((node) => node.field)).toEqual([
      'area',
      'province',
      'city',
      'type',
      'money',
      'number',
    ]);
    expect(hiddenColumnsDetail).toHaveLength(1);
    expect(measureDetail.displaySiblingNode.prev?.field).toEqual('type');
    expect(measureDetail.displaySiblingNode.next?.field).toEqual('number');
    expect(measureDetail.hideColumnNodes).toHaveLength(1);
    expect(measureDetail.hideColumnNodes[0].field).toEqual('price');
  });

  test('should render correctly resize group for custom column fields', async () => {
    s2.setTheme({
      resizeArea: {
        backgroundOpacity: 1,
      },
    });
    await s2.render(false);

    const groups = s2.facet.foregroundGroup.getElementById<Group>(
      KEY_GROUP_COL_RESIZE_AREA,
    );

    expect(groups?.childNodes.length).toEqual(11);
  });

  test('should format custom column', async () => {
    s2.setDataCfg({
      ...customColDataCfg,
      meta: [
        {
          field: 'city',
          formatter: (value) => `#-${value}`,
        },
        {
          field: 'price',
          formatter: () => '2',
        },
      ],
    });
    await s2.render();

    const { colNodes, dataCellTexts } = mapCellNodeValues(s2);

    expect(colNodes).toMatchSnapshot();
    expect(dataCellTexts).toMatchSnapshot();

    // 列头不应该按 formatter 格式化
    expect(
      colNodes.every((node) => node.actualText === node.value),
    ).toBeTruthy();
  });

  test('should render default series number text', async () => {
    s2.setOptions({
      seriesNumber: {
        enable: true,
      },
    });
    await s2.render(false);

    expect(s2.facet.getColNodes()[0].value).toEqual('序号');
  });

  test('should render custom series number text', async () => {
    const seriesNumberText = '牛';

    s2.setOptions({
      seriesNumber: {
        enable: true,
        text: seriesNumberText,
      },
    });
    await s2.render(false);

    expect(s2.facet.getColNodes()[0].value).toEqual(seriesNumberText);
  });

  test('should render correctly column height if enable series number', async () => {
    s2.setOptions({
      seriesNumber: {
        enable: true,
      },
    });
    await s2.render(false);

    const nodes = s2.facet
      .getColNodes()
      .filter((node) => node.level === 0 && node.isLeaf);

    const cellHeight = s2.facet.getLayoutResult().colsHierarchy.height;

    expect(cellHeight).toEqual(60);
    expect(nodes.every((node) => node.height === cellHeight)).toBeTruthy();
  });

  test('should not sample series node', async () => {
    s2.setOptions({
      seriesNumber: {
        enable: true,
      },
    });
    await s2.render(false);

    const { sampleNodeForLastLevel, sampleNodesForAllLevels } =
      s2.facet.getLayoutResult().colsHierarchy;

    expect(sampleNodeForLastLevel?.isSeriesNumberNode()).toBeFalsy();
    expect(
      sampleNodesForAllLevels.every((node) => !node.isSeriesNumberNode()),
    ).toBeTruthy();
  });

  test('should render custom multiple column nodes', async () => {
    s2.setDataCfg({
      ...baseDataConfig,
      fields: {
        columns: customColMultipleColumns,
      },
    });
    await s2.render();

    const colNodes = s2.facet.getColNodes().map((node) => {
      return {
        value: node.value,
        width: node.width,
        height: node.height,
        description: node.extra?.['description'],
      };
    });

    expect(colNodes).toHaveLength(6);
    expect(colNodes).toMatchSnapshot();
  });

  test('should render correctly resize area handler', async () => {
    s2.setDataCfg({
      ...baseDataConfig,
      fields: {
        columns: customColMultipleColumns,
      },
    });
    s2.setTheme({
      resizeArea: {
        backgroundOpacity: 1,
      },
    });
    await s2.render();

    const resizeArea = s2.facet.foregroundGroup.getElementById<Group>(
      KEY_GROUP_COL_RESIZE_AREA,
    )!;
    const resizeAreaList = resizeArea.children as CustomRect[];

    expect(resizeAreaList.length).toEqual(8);
  });
});
