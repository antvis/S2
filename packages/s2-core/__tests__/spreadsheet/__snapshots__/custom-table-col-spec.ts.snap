// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSheet Custom Tests should calc correctly col index of leaf nodes 1`] = `
Array [
  Object {
    "colIndex": 0,
    "value": "省份",
  },
  Object {
    "colIndex": 1,
    "value": "城市",
  },
  Object {
    "colIndex": 2,
    "value": "类型",
  },
  Object {
    "colIndex": 3,
    "value": "价格",
  },
  Object {
    "colIndex": 4,
    "value": "数量",
  },
]
`;

exports[`TableSheet Custom Tests should calc correctly leaf nodes width after column resized 1`] = `
Array [
  Object {
    "height": 100,
    "value": "地区",
  },
  Object {
    "height": 30,
    "value": "省份",
  },
  Object {
    "height": 30,
    "value": "城市",
  },
  Object {
    "height": 130,
    "value": "类型",
  },
  Object {
    "height": 100,
    "value": "金额",
  },
  Object {
    "height": 30,
    "value": "价格",
  },
  Object {
    "height": 30,
    "value": "数量",
  },
]
`;

exports[`TableSheet Custom Tests should format custom column 1`] = `
Array [
  Object {
    "actualText": "地区",
    "id": "area",
    "value": "地区",
  },
  Object {
    "actualText": "省份",
    "id": "province",
    "value": "省份",
  },
  Object {
    "actualText": "城市",
    "id": "city",
    "value": "城市",
  },
  Object {
    "actualText": "类型",
    "id": "type",
    "value": "类型",
  },
  Object {
    "actualText": "金额",
    "id": "money",
    "value": "金额",
  },
  Object {
    "actualText": "价格",
    "id": "price",
    "value": "价格",
  },
  Object {
    "actualText": "数量",
    "id": "number",
    "value": "数量",
  },
]
`;

exports[`TableSheet Custom Tests should format custom column 2`] = `
Array [
  "浙江省",
  "浙江省",
  "浙江省",
  "浙江省",
  "浙江省",
  "浙江省",
  "浙江省",
  "浙江省",
  "浙江省",
  "浙江省",
  "浙江省",
  "浙江省",
  "#-杭州市",
  "#-绍兴市",
  "#-宁波市",
  "#-舟山市",
  "#-杭州市",
  "#-绍兴市",
  "#-宁波市",
  "#-舟山市",
  "#-杭州市",
  "#-绍兴市",
  "#-宁波市",
  "#-舟山市",
  "家具",
  "家具",
  "家具",
  "家具",
  "家具",
  "家具",
  "家具",
  "家具",
  "办公用品",
  "办公用品",
  "办公用品",
  "办公用品",
  "2",
  "2",
  "2",
  "2",
  "2",
  "2",
  "2",
  "2",
  "2",
  "2",
  "2",
  "2",
  "7789",
  "2367",
  "3877",
  "4342",
  "5343",
  "632",
  "7234",
  "834",
  "945",
  "1304",
  "1145",
  "1432",
]
`;

exports[`TableSheet Custom Tests should render custom layout column nodes 1`] = `
Array [
  Object {
    "description": undefined,
    "height": 30,
    "value": "地区",
    "width": 239.6,
  },
  Object {
    "description": undefined,
    "height": 30,
    "value": "省份",
    "width": 119.8,
  },
  Object {
    "description": undefined,
    "height": 30,
    "value": "城市",
    "width": 119.8,
  },
  Object {
    "description": undefined,
    "height": 60,
    "value": "类型",
    "width": 119.8,
  },
  Object {
    "description": undefined,
    "height": 30,
    "value": "金额",
    "width": 239.6,
  },
  Object {
    "description": "价格描述",
    "height": 30,
    "value": "价格",
    "width": 119.8,
  },
  Object {
    "description": undefined,
    "height": 30,
    "value": "数量",
    "width": 119.8,
  },
]
`;

exports[`TableSheet Custom Tests should render custom multiple column nodes 1`] = `
Array [
  Object {
    "description": "a-1 描述",
    "height": 30,
    "value": "自定义节点 a-1",
    "width": 449.25,
  },
  Object {
    "description": "a-1-1 描述",
    "height": 30,
    "value": "自定义节点 a-1-1",
    "width": 299.5,
  },
  Object {
    "description": "指标1描述",
    "height": 30,
    "value": "指标1",
    "width": 149.75,
  },
  Object {
    "description": "指标2描述",
    "height": 30,
    "value": "指标2",
    "width": 149.75,
  },
  Object {
    "description": "a-1-2 描述",
    "height": 60,
    "value": "自定义节点 a-1-2",
    "width": 149.75,
  },
  Object {
    "description": "a-2 描述",
    "height": 90,
    "value": "自定义节点 a-2",
    "width": 149.75,
  },
]
`;
